src/ngraph/runtime/cpu/pass/cpu_memory_assignment.cpp

```diff
@@ -0,0 +1,886 @@
+        case CPUTensorRole::INTERMEDIATE: return string("CPUTensorRole::INTERMEDIATE");
+        case CPUTensorRole::CONSTANT: return string("CPUTensorRole::CONSTANT");
+        case CPUTensorRole::OUTPUT: return string("CPUTensorRole::OUTPUT");
+        }
+        throw runtime_error("unhandled CPU tensor role");
+    };
+
+    //liveness analysis
+    unordered_set<size_t> allocated_sets;
+    unordered_set<size_t> freed_sets;
+    NGRAPH_DEBUG << "cpu_memory_assignment: m_key_to_tensors_set_map:";
+    for (auto& ele : m_key_to_tensors_set_map)
+    {
+        NGRAPH_DEBUG << "key : " << ele.first << "; label: " << find_role(ele.second.first)
+                     << "; sets: {";
+        for (auto& ele_t : ele.second.second)
+        {
+            NGRAPH_DEBUG << ele_t->get_name() << " ";
+        }
+        NGRAPH_DEBUG << "}";
+        if (ele.second.first != CPUTensorRole::INTERMEDIATE)
+        {
+            // do not allocate and free memory for function inputs, outputs, constants and tensors
+            // sharing memory with them.
+            allocated_sets.insert(ele.first);
+            freed_sets.insert(ele.first);
+        }
+    }
+
+    // forward pass to build liveness_new_list
+    for (auto it = ops.begin(); it != ops.end(); it++)
+    {
+        const shared_ptr<Node>& node = *it;
+        node->liveness_new_list.clear();
+
+        for (size_t i = 0; i < node->get_output_size(); ++i)
+        {
+            auto tensor = &node->get_output_tensor(i);
+            auto key = m_tensor_to_key_map[tensor];
+            if (allocated_sets.find(key) == allocated_sets.end())
+            {
+                node->liveness_new_list.insert(tensor);
+                allocated_sets.insert(key);
+            }
+        }
+    }
+
+    // backward pass to build liveness_free_list
+    for (auto it = ops.rbegin(); it != ops.rend(); it++)
+    {
+        const shared_ptr<Node>& node = *it;
+        node->liveness_free_list.clear();
+
+        for (descriptor::Input& input_decl : node->get_inputs())
+        {
+            auto tensor = &input_decl.get_tensor();
+            auto key = m_tensor_to_key_map[tensor];
+            if (freed_sets.find(key) == freed_sets.end())
+            {
+                node->liveness_free_list.insert(tensor);
+                freed_sets.insert(key);
+            }
+        }
+    }
+
+    // memory assignment using liveness analysis result
+
+    // memory manager for non-cacheable ops, memory allocation will be freed when not longer in use
+    ngraph::pass::MemoryManager mm(m_alignment, m_disable_memory_sharing);
+    // memory manager for cacheable ops, memory allocation will never be freed
+    ngraph::pass::MemoryManager mm_caching(m_alignment, true);
+
+    // reuse memory
+    if (!m_disable_memory_sharing)
+    {
+        // build caching map from cacheability
+        for (shared_ptr<Node> node : function->get_ordered_ops())
```

Comment:
I think `node` should be a reference, so avoid the ++ and -- of the counter.

Reply:
what counter?